File: 2026/3o-termo/programacao-back-end/demonstracoes/apis/json/manipulando/js/granularidade/crud.js
const fs = require("fs");
const prompt = require("prompt-sync")();

// Menu principal
function menu() {
  console.log("\n=== MENU CURSO ===");
  console.log("1 - Adicionar contato");
  console.log("2 - Listar contatos");
  console.log("3 - Atualizar contato");
  console.log("4 - Excluir contato");
  console.log("X - Sair");
}

function escolherGrupo() {
  console.log("Tipo de contato: ");
  console.log("1. Aluno");
  console.log("2. Professor");

  const opcao = prompt("Escolha: ");

  if (opcao === "1") return "alunos";
  if (opcao === "2") return "professores";

  console.log("Opção inválida.");
  return null;
}

// Leitura do JSON
function lerDados() {
  const dados = fs.readFileSync("pessoasCurso.json", "utf-8");
  return JSON.parse(dados);
}

function salvarDados(dados) {
  fs.writeFileSync("pessoasCurso.json", JSON.stringify(dados, null, 2));
}

function adicionar() {
  const grupo = escolherGrupo();
  if (!grupo) return;

  const nome = prompt("Nome: ");
  const telefone = prompt("Telefone: ");

  const dados = lerDados();
  dados[grupo].push({ nome, telefone });

  salvarDados(dados);
  console.log("Contato adicionado com sucesso!");
}

function listar() {
  const grupo = escolherGrupo();
  if (!grupo) return;

  const dados = lerDados();

  console.log(`\nLista de ${grupo.toUpperCase()}: `);
  dados[grupo].forEach((contato, index) => {
    console.log(`${index + 1}. ${contato.nome} - ${contato.telefone}`);
  });
}

function atualizar() {
  const grupo = escolherGrupo();
  if (!grupo) return;

  const dados = lerDados();
  const index =
    parseInt(prompt("Digite o ID do contato que deseja atualizar: ")) - 1;

  if (index < 0 || index >= dados[grupo].length || isNaN(index)) {
    console.log("ID inválido.");
    return;
  }

  const nome = prompt("Novo nome: ");
  const telefone = prompt("Novo telefone: ");

  dados[grupo][index] = { nome, telefone };
  salvarDados(dados);
  console.log("Contato atualizado com sucesso!");
}

function excluir() {
  const grupo = escolherGrupo();
  if (!grupo) return;

  const dados = lerDados();
  const index =
    parseInt(prompt("Digite o ID do contato que deseja excluir: ")) - 1;

  if (index < 0 || index >= dados[grupo].length || isNaN(index)) {
    console.log("ID inválido.");
    return;
  }

  dados[grupo].splice(index, 1);
  salvarDados(dados);
  console.log("Contato excluído com sucesso!");
}

function main() {
  let opcao;

  do {
    menu();
    opcao = prompt("Escolha uma opção: ");
    switch (opcao) {
      case "1":
        adicionar();
        break;
      case "2":
        listar();
        break;
      case "3":
        atualizar();
        break;
      case "4":
        excluir();
        break;
      case "X":
      case "x":
        console.log("Alunos! Adeus!");
        return;
      default:
        console.log("Digite alguma das opções!");
        break;
    }
    prompt("Digite enter para continuar... ");
    console.clear();
  } while (opcao.toUpperCase() !== "X");
}

main();

File: 2026/3o-termo/programacao-backend/demonstracoes/apis/json/manipulando/py/granularidade/crud.py
import json
import os

# Menu Principal

def menu():
    print("\n === Menu de Contatos ===")
    print("1. Adicionar")
    print("2. Listar")
    print("3. Atualizar")
    print("4. Excluir")
    print("5. Sair")

# Função para escolher o grupo

def escolher_grupo():
    print("\nTipo de contato")
    print("1. Aluno")
    print("2. Professor")

    opcao = input("Escolha: ")

    match (opcao):
        case "1":
            return "alunos"
        case "2":
            return "professores"
        case _:
            print("Opção inválida.")
            return None

# Função para ler o JSON
def ler_dados():
    with open("contatos.json", "r", encoding="utf-8") as arquivo:
        return json.load(arquivo)

def salvar_dados(dados):
    with open("contatos.json", "w", encoding="utf-8") as arquivo:
        json.dump(dados, arquivo, indent=4, ensure_ascii=False)

# Adicionar dados
def adicionar():
    grupo = escolher_grupo()

    if not grupo: 
        return
    
    nome = input("Nome: ")
    telefone = input("Telefone: ")

    dados = ler_dados()

    dados[grupo].append({
        "nome": nome,
        "telefone": telefone
    })

    salvar_dados(dados)
    print("Contato adicionado com sucesso!")

# Função listar contatos
def listar():
    grupo = escolher_grupo()

    if not grupo:
        return

    dados = ler_dados()
    print(f"\nLista de {grupo.title()}")

    for index, contato in enumerate(dados[grupo], start=1):
        print(f"{index}. {contato['nome']} - {contato['telefone']}")

# Atualizar
def atualizar():
    grupo = escolher_grupo()

    if not grupo:
        return
    
    dados = ler_dados()

    index = int(input("Index do contato: ")) -1

    if index < 0 or index > len(dados[grupo]):
        print("ID inválido!")
        return

    nome = input("Novo nome:")
    telefone = input("Novo telefone: ")

    dados[grupo][index] = {
        "nome": nome,
        "telefone": telefone
    }

    salvar_dados(dados)
    print("Contato atualizado!")

def excluir():
    grupo = escolher_grupo()
    if not grupo:
        return
    
    dados = ler_dados()

    index = int(input("Índice do contato: ")) - 1

    # Verificar se o Index é valido
    
    if index < 0 or index > len(dados[grupo]):
        print("ID inválido!")
        return
    
    dados[grupo].pop(index)
    salvar_dados()
    print("Contato excluído com sucesso.")


def main():

    while True:
        menu()
        opcao = input("\nEscolha: ")
        
        match(opcao):
            case "1":
                adicionar()
            case "2":
                listar()
            case "3":
                atualizar()
            case "4":
                excluir
            case "5":
                print("Até mais!")
                break
            case _:
                print("Opção inválida!")
            
        input("Digite ENTER para continuar.\n")
        os.system('cls' if os.name == "nt" else 'clear')
